### PR TITLE
[dhcp_server] Add show dhcp_server ipv4 lease

### PR DESCRIPTION
**Pull request by Xichen96**
#### Why I did it
Add dhcp server cli show dhcp_server ipv4 lease

##### Work item tracking
- Microsoft ADO **(number only)**:

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211
- [ ] 202305

#### Tested branch (Please provide the tested image version)

master image from https://github.com/sonic-net/sonic-buildimage/pull/17051

#### Description for the changelog

Add show dhcp_server ipv4 lease.

### CODE REVIEW

File: dockers/docker-dhcp-server/cli/show/plugins/show_dhcp_server.py [hunk 1]
```diff
@@ -1,13 +1,46 @@
...
+        if len(items) != 3:
+            continue
+        if items[0] != "DHCP_SERVER_IPV4_LEASE":
+            continue
+        if dhcp_interface and items[1] != dhcp_interface:
+            continue
+        entry = dbconn.get_all("STATE_DB", key)
+        interface = items[1]
+        mac = items[2]
+        port = dbconn.get("STATE_DB", "FDB_TABLE|" + interface + ":" + mac, "port")
```
Comment (Blueve) [inline dockers/docker-dhcp-server/cli/show/plugins/show_dhcp_server.py:40]: What if the mac address cannot be retrieved from FDB?

File: dockers/docker-dhcp-server/cli/show/plugins/show_dhcp_server.py [hunk 2]
```diff
@@ -1,13 +1,46 @@
...
+        table.append([interface + "|" + port, mac, entry["ip"], entry["lease_start"], entry["lease_end"]])
+    print(tabulate(table, headers=headers))
```
Comment (Blueve) [inline dockers/docker-dhcp-server/cli/show/plugins/show_dhcp_server.py:42]: Why not use `click.echo()`?

File: dockers/docker-dhcp-server/cli/show/plugins/show_dhcp_server.py (hunk 1 above)
Comment (Blueve) [reply dockers/docker-dhcp-server/cli/show/plugins/show_dhcp_server.py:40]: Display '<Unknown>' if port not found in FDB table

File: dockers/docker-dhcp-server/cli-plugin-tests/conftest.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
+import pytest
+import mock_tables # lgtm [py/unused-import]
+from unittest import mock
+
+@pytest.fixture()
+def mock_db():
+    db = mock.Mock()
+    mock_data = {
+        "STATE_DB": {
```
Comment (Blueve) [inline dockers/docker-dhcp-server/cli-plugin-tests/conftest.py:9]: mock_config_db.json
mock_state_db.json

File: dockers/docker-dhcp-server/cli-plugin-tests/test_show_dhcp_server.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    def test_plugin_registration(self):
+        cli = mock.MagicMock()
+        show_dhcp_server.register(cli)
+
+    def test_show_dhcp_server_ipv4_lease(self, mock_db):
+        runner = CliRunner()
+        db = Db()
+        db.db = mock_db()
+        result = runner.invoke(show_dhcp_server.dhcp_server.ipv4.lease, [])
+        assert result.exit_code == 0, "exit code: {}, Exception: {}, Traceback: {}".format(result.exit_code, result.exception, result.exc_info)
```
Comment (Blueve) [inline dockers/docker-dhcp-server/cli-plugin-tests/test_show_dhcp_server.py:20]: check cmd output

File: dockers/docker-dhcp-server/cli/show/plugins/show_dhcp_server.py [hunk 5]
```diff
@@ -1,13 +1,46 @@
...
+
+
+@ipv4.command()
+@click.argument('dhcp_interface', required=False)
+@clicommon.pass_db
+def lease(db, dhcp_interface):
+    headers = ["Interface", "MAC Address", "IP", "Lease Start", "Lease End"]
+    table = []
+    dbconn = db.db
+    for key in dbconn.keys("STATE_DB"):
```
Comment (Blueve) [inline dockers/docker-dhcp-server/cli/show/plugins/show_dhcp_server.py:29]: Try use `db.keys` method

File: dockers/docker-dhcp-server/cli-plugin-tests/test_show_dhcp_server.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+sys.path.append('../cli/show/plugins/')
+import show_dhcp_server
+
+
+class TestShowDHCPServer(object):
+    def test_plugin_registration(self):
+        cli = mock.MagicMock()
+        show_dhcp_server.register(cli)
+
+    def test_show_dhcp_server_ipv4_lease(self, mock_db):
```
Comment (Blueve) [inline dockers/docker-dhcp-server/cli-plugin-tests/test_show_dhcp_server.py:15]: test_show_dhcp_server_ipv4_lease_with_dhcpintf
test_show_dhcp_server_ipv4_lease_without_dhcpintf
test_show_dhcp_server_ipv4_lease_client_not_in_fdb

File: dockers/docker-dhcp-server/cli-plugin-tests/test_show_dhcp_server.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+class TestShowDHCPServer(object):
+    def test_plugin_registration(self):
+        cli = mock.MagicMock()
+        show_dhcp_server.register(cli)
+
+    def test_show_dhcp_server_ipv4_lease_without_dhcpintf(self, mock_db):
+        expected_stdout = """\
+Interface            MAC Address        IP             Lease Start    Lease End
+-------------------  -----------------  -----------  -------------  -----------
+Vlan1000|Ethernet10  10:70:fd:b6:13:00  192.168.0.1     1677640581   1677641481
```
Comment (Blueve) [inline dockers/docker-dhcp-server/cli-plugin-tests/test_show_dhcp_server.py:21]: We need convert the timestamp to datetime format. @yaqiangz for fyi

File: dockers/docker-dhcp-server/cli/config/plugins/dhcp_server.py [hunk 8]
```diff
@@ -1,17 +1,309 @@
...
 
-def register(cli):
-    # cli.add_command(dhcp_server)
+@dhcp_server.group(cls=clicommon.AliasedGroup, name="ipv4")
+def dhcp_server_ipv4():
+    """Show ipv4 related dhcp_server info"""
+    pass
+
+
+def validate_str_type(type, value):
```
Comment (Blueve) [inline dockers/docker-dhcp-server/cli/config/plugins/dhcp_server.py:19]: Where you use this method?

**Review by yaqiangz — APPROVED** [review]

**Review by Blueve — APPROVED** [review]

File: dockers/docker-dhcp-server/cli-plugin-tests/test_show_dhcp_server.py (hunk 7 above)
Comment (yaqiangz) [reply dockers/docker-dhcp-server/cli-plugin-tests/test_show_dhcp_server.py]: Yes, we need to convert it to datetime

File: dockers/docker-dhcp-server/cli/show/plugins/show_dhcp_server.py (hunk 1 above)
Comment (Blueve) [reply dockers/docker-dhcp-server/cli/show/plugins/show_dhcp_server.py:38]: Closed

File: dockers/docker-dhcp-server/cli/show/plugins/show_dhcp_server.py (hunk 2 above)
Comment (Blueve) [reply dockers/docker-dhcp-server/cli/show/plugins/show_dhcp_server.py]: Closed

File: dockers/docker-dhcp-server/cli-plugin-tests/conftest.py (hunk 3 above)
Comment (Blueve) [reply dockers/docker-dhcp-server/cli-plugin-tests/conftest.py]: Closed

File: dockers/docker-dhcp-server/cli-plugin-tests/test_show_dhcp_server.py (hunk 4 above)
Comment (Blueve) [reply dockers/docker-dhcp-server/cli-plugin-tests/test_show_dhcp_server.py:56]: Closed

File: dockers/docker-dhcp-server/cli/show/plugins/show_dhcp_server.py (hunk 5 above)
Comment (Blueve) [reply dockers/docker-dhcp-server/cli/show/plugins/show_dhcp_server.py]: Closed

File: dockers/docker-dhcp-server/cli-plugin-tests/test_show_dhcp_server.py (hunk 6 above)
Comment (Blueve) [reply dockers/docker-dhcp-server/cli-plugin-tests/test_show_dhcp_server.py]: Closed

File: dockers/docker-dhcp-server/cli-plugin-tests/test_show_dhcp_server.py (hunk 7 above)
Comment (Blueve) [reply dockers/docker-dhcp-server/cli-plugin-tests/test_show_dhcp_server.py]: Closed

File: dockers/docker-dhcp-server/cli/config/plugins/dhcp_server.py (hunk 8 above)
Comment (Blueve) [reply dockers/docker-dhcp-server/cli/config/plugins/dhcp_server.py]: Closed

**Comment by Blueve** [comment]
@yxieca could you help merge this PR?

**Pull request merged**